ifsbench/job.py

```diff
@@ -106,206 +90,134 @@ class CpuDistribution(Enum):
-        self.cpu_config = cpu_config
-        if tasks is not None:
-            self.tasks = tasks
-        if nodes is not None:
-            self.nodes = nodes
-        if tasks_per_node is not None:
-            self.tasks_per_node = tasks_per_node
-        if tasks_per_socket is not None:
-            self.tasks_per_socket = tasks_per_socket
-        if cpus_per_task is not None:
-            self.cpus_per_task = cpus_per_task
-        if threads_per_core is not None:
-            self.threads_per_core = threads_per_core
-        if gpus_per_task is not None:
-            self.gpus_per_task = gpus_per_task
-        if bind is not None:
-            self.bind = bind
-        if distribute_remote is not None:
-            self.distribute_remote = distribute_remote
-        if distribute_local is not None:
-            self.distribute_local = distribute_local
-
-
-        gpus_per_node = self.get_gpus_per_task() * self.get_tasks_per_node()
-        if gpus_per_node > self.cpu_config.gpus_per_node:
-            raise ValueError("More GPUs have been requested than are available.")
-
-        try:
-            tasks = self.get_tasks()
-            nodes = self.get_nodes()
-            threads = self.get_threads()
-        except AttributeError as excinfo:
-            error(('Need to specify at least one of:\n'
-                   'number of tasks or (tasks_per_node and nodes) or (tasks_per_socket and nodes)'))
-            raise excinfo
-
-        if nodes < 1:
-            error(f'Invalid number of nodes: {nodes}')
-            raise ValueError
-        if tasks < 1:
-            error(f'Invalid number of tasks: {nodes}')
-            raise ValueError
-        if threads < tasks:
-            error(f'Invalid number of threads: {threads}')
-            raise ValueError
-
-    def get_tasks(self):
-        """
-        The total number of MPI tasks
+    tasks : int = None
+    nodes : int = None
+    tasks_per_node : int = None
+    tasks_per_socket : int = None
+    cpus_per_task : int = None
+    threads_per_core : int = None
+    gpus_per_task : int = None
+    account : str = None
+    partition : str = None
+    bind : CpuBinding = None
+    distribute_remote : CpuDistribution = None
+    distribute_local : CpuDistribution = None
 
-        If this has not been specified explicitly, it is estimated as
-        ``nodes * tasks_per_node``.
+    def copy(self):
         """
-        tasks = getattr(self, 'tasks', None)
-        if tasks is None:
-            return self.get_nodes() * self.get_tasks_per_node()
-        return tasks
-
-    def get_nodes(self):
+        Return a deep copy of this object.
         """
-        The total number of nodes
 
-        If this has not been specified explicitly, it is estimated as
-        ``ceil(threads / available_threads_per_node)`` with the number of
-        available threads dependent on the use of SMT.
+        return replace(self)
+
+
+
+    def calculate_missing(self, cpu_configuration):
```

Comment:
If we used pydantic, this could maybe be implemented via field_validators or similar